text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+  For instance, in the presence of algebraic types, there is a specific incarnation of projection steps involving constructors. For instance, the problem `Γ ⊢ match ?e with S n => t | O => u end =?= v` can be solved (among other resolution strategies) by defining `?e := S ?n` for a new existential variable `?n` and solving `Γ ⊢ t[n:=?n] =?= v` or defining `?e := 0` and solving `Γ ⊢ u =?= v`.
+
+  We call projection of variables the subcase when `u` is a variable and `E` is empty. In practice, Coq implements only projections of variables and projections of terms of the form `C t₁ ... tₙ` for `C` a constructor, with `E` empty (but PR [#14798](https://github.com/coq/coq/pull/14798) extends this to `E` made of nesting of `match` and projections).
+
+- *Imitation*, that is, imitating the structure of `t`, or part of it, and generating subproblems about the subterms of the imitated part of `t`. 
+
+  For instance, if `C` is a constructor, the equation `Γ ⊢ ?e@{subst} =?= C t₁ ... tₙ` reduces to the collection of equations `Γ ⊢ ?eᵢ@{subst} =?= tᵢ` by declaring new existential variables `?eᵢ` of appropriate type in the same context as `?e` and by setting `?e := C ?e₁@{dom(Δ)} ... ?eₙ@{dom(Δ)}`. 
+
+  Similarly, if the head symbol of `t` is a binder, new existential variables are generated for the subterms of the binder in a context extended with the new binder.
+
+## Efficiency non-determinism
+
+The non-determinism comes from the potentially different ways to project and from the ability to also sometimes imitate. For instance, there are 2 possible projections and one possible imitation to resolve `Γ ⊢ ?e@{a:=t,b:=t} =?= t` for `t` variable-free.
+
+There is also a form of non-determinism which comes from reducing unification problems or not in normal form:
+- When resolving `Γ ⊢ c t =?= c' t'` with `c` and `c'` distinct unfoldable constants, there is a choice in reducing `c` or `c'` in the hope that the problem is eventually found satisfiable. This choice can lead to arbitrarily long computations so, controlling unfolding is important (there are similar problems when efficiently comparing the convertibility of two terms).
+- It is also useful to limit unfolding of constants so as to obtain solutions as close as possible to the original form of the terms involved.
+
+## Higher order pattern unification
+
+*Higher order pattern unification* (see Miller 1991 and Pfenning 1991) is a special case of simple unification problem when the substitution is made only of distinct variables. In this case, there is a single solution obtained by projecting the unique possible variable or imitating if not a variable.
+
+Pattern unification has various refinements (see e.g. Libal and Miller 2016) such as:
+- expecting distinct variables only among the variables that actually occur in the term to match
```

Comment:
Note that this can cause the pattern solution to be incorrect if there are other variables in the context which are assigned to terms which are headed by a constructor, because we may choose whether or not to match on them.  (Consider `x, y : bool |- ?e` and `x : bool |- ?e@{x:=x,y:=true} == x`, we can solve this with either `x` or with `if y then x else _`, and the latter may be manded by a future unification problem.)

Reply:
This should be formulated clearly. It is `among the variables`, so the presence of `true` makes that it is not a pattern (see the `Option.List.map` in `Evarsolve.is_unification_pattern_evar`; the restriction to actual dependencies in `get_actual_deps` comes after).